// pg-minify/lib/utils.js
'use strict';

const util = require('util');

///////////////////////////////////////////////////////
// Returns {line, column} of an index within the text.
function getIndexPos(text, index) {
    let lineIdx = 0, colIdx = index, pos = 0;
    do {
        pos = text.indexOf('\n', pos);
        if (pos === -1 || index < pos + 1) {
            break;
        }
        lineIdx++;
        pos += 1;
        colIdx = index - pos;
    } while (pos < index);
    return {
        line: lineIdx + 1,
        column: colIdx + 1
    };
}

///////////////////////////////////////////
// Returns a space gap for console output.
function messageGap(level) {
    return ' '.repeat(level * 4);
}

////////////////////////////////////////////////////
// Legacy type inspection, to support Node.js < 6.x
function addInspection(type, cb) {
    // istanbul ignore else;
    if (util.inspect.custom) {
        type.prototype[util.inspect.custom] = cb;
    } else {
        type.prototype.inspect = cb;
    }
}

module.exports = {
    getIndexPos,
    messageGap,
    addInspection
};
