// dotnet-deps-parser/dist/parsers/project-assets-json-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
function getProjectNameForProjectAssetsJson(manifestFile) {
    return _.get(manifestFile, ['project', 'restore', 'projectName'], {});
}
function getProjectVersionForProjectAssetsJson(manifestFile) {
    return _.get(manifestFile, ['project', 'version'], {});
}
function buildPackageTree(name, version) {
    var depTree = {
        dependencies: {},
        name: name,
        version: version,
    };
    return depTree;
}
// Currently the function getDependencyTreeFromProjectAssetsJson returns
// a two level deep flat list of 100% of dependencies.
// TODO: Get full tree
function getDependencyTreeFromProjectAssetsJson(manifestFile, targetFrameWork) {
    var projectName = getProjectNameForProjectAssetsJson(manifestFile);
    var projectVersion = getProjectVersionForProjectAssetsJson(manifestFile);
    var depTree = buildPackageTree(projectName, projectVersion);
    var topLevelDeps = Object.keys(_.get(manifestFile, ['targets', targetFrameWork], {}));
    for (var _i = 0, topLevelDeps_1 = topLevelDeps; _i < topLevelDeps_1.length; _i++) {
        var topLevelDep = topLevelDeps_1[_i];
        var _a = topLevelDep.split('/'), topLevelDepName = _a[0], topLevelDepVersion = _a[1];
        var topLevelDepTree = buildPackageTree(topLevelDepName, topLevelDepVersion);
        var transitiveDeps = _.get(manifestFile, ['targets', targetFrameWork, topLevelDep, 'dependencies'], {});
        for (var _b = 0, _c = Object.keys(transitiveDeps); _b < _c.length; _b++) {
            var transitiveDep = _c[_b];
            var transitiveDepVersion = transitiveDeps[transitiveDep];
            var transitiveDepTree = buildPackageTree(transitiveDep, transitiveDepVersion);
            topLevelDepTree.dependencies[transitiveDep] = transitiveDepTree;
        }
        depTree.dependencies[topLevelDepName] = topLevelDepTree;
    }
    return depTree;
}
exports.getDependencyTreeFromProjectAssetsJson = getDependencyTreeFromProjectAssetsJson;
