// @snyk/dep-graph/dist/legacy/event-loop-spinner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var EventLoopSpinner = /** @class */ (function () {
    function EventLoopSpinner(thresholdMs) {
        if (thresholdMs === void 0) { thresholdMs = 10; }
        this.thresholdMs = thresholdMs;
        this.afterLastSpin = Date.now();
    }
    EventLoopSpinner.prototype.isStarving = function () {
        return (Date.now() - this.afterLastSpin) > this.thresholdMs;
    };
    EventLoopSpinner.prototype.reset = function () {
        this.afterLastSpin = Date.now();
    };
    EventLoopSpinner.prototype.spin = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve) { return setImmediate(function () {
                        _this.reset();
                        resolve();
                    }); })];
            });
        });
    };
    return EventLoopSpinner;
}());
exports.EventLoopSpinner = EventLoopSpinner;
