// @hapist/whitelist/dist/config/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("@hapi/joi"));
const lodash_get_1 = __importDefault(require("lodash.get"));
const lodash_set_1 = __importDefault(require("lodash.set"));
const schema_1 = require("./schema");
class Config {
    constructor() {
        this.config = {};
    }
    load(options) {
        this.reset();
        const { value, error } = joi_1.default.validate(options, schema_1.schema);
        if (error) {
            this.error = error;
            return;
        }
        this.config = value;
    }
    reset() {
        this.error = undefined;
        this.config = {};
    }
    get(key, defaultValue) {
        return lodash_get_1.default(this.config, key, defaultValue);
    }
    set(key, value) {
        lodash_set_1.default(this.config, key, value);
    }
    getError() {
        return this.error;
    }
    hasError() {
        return !!this.error;
    }
}
exports.config = new Config();
