// log-process-errors/build/src/level.js
"use strict";require("core-js/modules/es.array.iterator");Object.defineProperty(exports,"__esModule",{value:true});exports.DEFAULT_LEVEL=exports.validateLevels=exports.getExampleLevels=exports.applyDefaultLevels=exports.getLevel=void 0;var _process=require("process");

var _jestValidate=require("jest-validate");

var _utils=require("./utils.js");


const getLevel=function({opts,name,error}){
const level=(0,_utils.result)(opts.level[name],error);

if(level==="default"||level===undefined){
return DEFAULT_LEVEL[name];
}

if(LEVELS.includes(level)){
return level;
}

(0,_process.emitWarning)(
`Invalid option 'level.${name}' returning '${level}': function must return one of ${LEVELS.join(
", ")
}`);


return DEFAULT_LEVEL[name];
};exports.getLevel=getLevel;


const applyDefaultLevels=function({
opts:{level:{default:defaultLevel,...level}={}}})
{
const levelA=(0,_utils.pickBy)(level,value=>value!==undefined);

if(defaultLevel===undefined){
return{...DEFAULT_LEVEL,...levelA};
}

const defaultLevels=(0,_utils.mapValues)(DEFAULT_LEVEL,()=>defaultLevel);
return{...defaultLevels,...levelA};
};exports.applyDefaultLevels=applyDefaultLevels;


const getExampleLevels=function(){
return(0,_utils.mapValues)(DEFAULT_LEVEL,getExampleLevel);
};exports.getExampleLevels=getExampleLevels;

const getExampleLevel=function(level){

return(0,_jestValidate.multipleValidOptions)(level,()=>{});
};

const validateLevels=function({level}){
Object.entries(level).forEach(validateLevel);
};exports.validateLevels=validateLevels;

const validateLevel=function([name,level]){
if(isValidLevel({level})){
return;
}

throw new Error(
`Invalid option 'level.${name}' '${level}': must be a function or one of ${LEVELS.join(
", ")
}`);

};

const isValidLevel=function({level}){
return(
LEVELS.includes(level)||level===undefined||typeof level==="function");

};

const LEVELS=["debug","info","warn","error","silent","default"];

const DEFAULT_LEVEL={
default:"error",
uncaughtException:"error",
warning:"warn",
unhandledRejection:"error",
rejectionHandled:"error",
multipleResolves:"info"};exports.DEFAULT_LEVEL=DEFAULT_LEVEL;
